pmd-java/src/main/resources/category/java/bestpractices.xml

```diff
@@ -802,6 +802,76 @@ public class MyTest {
         </example>
     </rule>
 
+    <rule name="JUnit5TestShouldBePackagePrivate"
+          language="java"
+          since="6.35"
+          message="JUnit 5 tests that use @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest should be package private."
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          typeResolution="true"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_bestpractices.html#junit5testshouldbepackageprivate">
+        <description>
+            JUnit 5 tests should be package private at the class level and for each method that uses @Test, @RepeatedTest, @TestFactory, @TestTemplate or @ParameterizedTest.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="version" value="2.0"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//ClassOrInterfaceDeclaration[
```

Comment:
In XPath 2 you can merge both expressions, like

```ruby
//ClassOrInterfaceDeclaration[...]/(
  self::*[@Abstract=false() and @PackagePrivate=false()]
| ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration/...
)
```

Please do this, as much of the rule is duplicated and it's hard to see what's going on. 

Also please report the violation on the MethodDeclaration and not the ClassOrInterfaceBodyDeclaration. Otherwise the line number of the violation is on the annotation...

Reply:
While I managed to report the violation on MethodDeclaration, I don't see how to reduce the two paths in one:
* the first one reports one violation on the class (if not package-private nor abstract) for any number of `@Test` methods contained in the class
* the second one reports one violation per `@Test` methods (if not package-private)

Could you elaborate? I'm probably missing something...

Reply:
While I managed to report the violation on MethodDeclaration, I don't see how to reduce the two paths in one:

    the first one reports one violation on the class (if not package-private nor abstract) for any number of @Test methods contained in the class
    the second one reports one violation per @Test methods (if not package-private)

Could you elaborate? I'm probably missing something...

Reply:
This is roughly what I meant:
```ruby
//ClassOrInterfaceDeclaration[
      (: a Junit 5 test class, ie, it has methods with the annotation :)
      ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration/Annotation[pmd-java:typeIs(...)]
]/(
    self::*[@Abstract=false() and @PackagePrivate=false()] 
|   ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration[Annotation[pmd-java:typeIs(...)]]
                        /MethodDeclaration[@PackagePrivate=false()]
)
```

So there's still two paths, but with a common ancestor.

Reply:
Thanks, I think I got it, please look at the updated rule.